Tasks/Gulp/gulptask.ts

```diff
@@ -2,118 +2,152 @@
+			}
+			else {
+				console.log(tl.loc('GulpReturnCode', code));
+			}
+		}, function (err) {
+			publishTestResults(publishJUnitResults, testResultsFiles);
 			tl.debug('taskRunner fail');
-			tl.setResult(tl.TaskResult.Failed, tl.loc('NpmFailed', err.message));
+			throw new Error(tl.loc('GulpFailed', err.message));
 		})
-	} else {
-		tl.setResult(tl.TaskResult.Succeeded, tl.loc('GulpReturnCode', code));
 	}
-}).fail(function (err) {
-	publishTestResults(publishJUnitResults, testResultsFiles);
-	tl.debug('taskRunner fail');
-	tl.setResult(tl.TaskResult.Failed, tl.loc('GulpFailed', err.message));
-})
+}
 
 function publishTestResults(publishJUnitResults, testResultsFiles: string) {
-    if (publishJUnitResults) {
-        //check for pattern in testResultsFiles
-        if (testResultsFiles.indexOf('*') >= 0 || testResultsFiles.indexOf('?') >= 0) {
-            tl.debug('Pattern found in testResultsFiles parameter');
-            var buildFolder = tl.getVariable('System.DefaultWorkingDirectory');
-            var allFiles = tl.find(buildFolder);
-            var matchingTestResultsFiles = tl.match(allFiles, testResultsFiles, { matchBase: true });
-        }
-        else {
-            tl.debug('No pattern found in testResultsFiles parameter');
-            var matchingTestResultsFiles = [testResultsFiles];
-        }
-        if (!matchingTestResultsFiles || matchingTestResultsFiles.length == 0) {
-            tl.warning('No test result files matching ' + testResultsFiles + ' were found, so publishing JUnit test results is being skipped.');
-            return 0;
-        }
-        var tp = new tl.TestPublisher("JUnit");
+	if (publishJUnitResults) {
+		//check for pattern in testResultsFiles
+		if (/[\*?+@!{}\[\]]/.test(testResultsFiles)) {
+			tl.debug('Pattern found in testResultsFiles parameter');
+			var matchingTestResultsFiles = tl.findMatch(tl.getVariable('System.DefaultWorkingDirectory'), testResultsFiles, null, { matchBase: true });
+		}
+		else {
+			tl.debug('No pattern found in testResultsFiles parameter');
+			var matchingTestResultsFiles = [testResultsFiles];
+		}
+		if (!matchingTestResultsFiles || matchingTestResultsFiles.length == 0) {
+			tl.warning('No test result files matching ' + testResultsFiles + ' were found, so publishing JUnit test results is being skipped.');
+			return 0;
+		}
+		var tp = new tl.TestPublisher("JUnit");
 		try {
 			tp.publish(matchingTestResultsFiles, true, "", "", "", true);
 		} catch (error) {
 			tl.warning(error);
 		}
-    }
+	}
 }
 
-function publishCodeCoverage(summaryFile) {
+function publishCodeCoverage(summaryFile, reportDirectory) {
 	try {
 		var ccPublisher = new tl.CodeCoveragePublisher();
 		ccPublisher.publish('cobertura', summaryFile, reportDirectory, "");
 	} catch (error) {
 		tl.warning(error);
 		throw error;
 	}
-}
\ No newline at end of file
+}
+
+function getGulpFiles(gulpFilePath: string): string[] {
+	var gulpFiles = [];
+
+	if (/[\*?+@!{}\[\]]/.test(gulpFilePath)) {
```

Comment:
I would recommend not pre-testing for special chars. The pattern should be: task authors can simply call tl.findMatch and it just do the right thing. There is already logic baked into the find functions to drill down as much as possible to avoid unnecessarily scanning a directory at a higher level.